Issue: upgrading from 3.7.0 to 3.8.0 breaks existing gitlab_project

### Terraform Version
1.1.2

### Affected Resource(s)
- gitlab_project

### Terraform Configuration Files
```hcl
resource "gitlab_project" "my_project" {
  for_each = var.projects
  name = each.key
  namespace_id = local.group_id_map[each.value.group]
  description = "Maintained by Terraform - UI changes will be lost."
  tags = each.value.tags
  container_registry_enabled = false
  request_access_enabled = true
  issues_enabled = false 
  merge_requests_enabled = true
  packages_enabled = false
  wiki_enabled = false
  pipelines_enabled = true
  build_coverage_regex = each.value.coverage
  approvals_before_merge = 1
  only_allow_merge_if_pipeline_succeeds = true
  only_allow_merge_if_all_discussions_are_resolved = true
  allow_merge_on_skipped_pipeline = false
  remove_source_branch_after_merge = true
  shared_runners_enabled = true
  push_rules {
    branch_name_regex = "main|(features|bugfixes)/${var.jira_slug}-.*"
    prevent_secrets   = true
    max_file_size     = 30
    member_check      = true
  }
}
```

### Debug Output
It is over 16K lines and contains so much sensitive material that I don't trust myself to find and redact every bit of it. Here are the highlights:
```
2022-01-03T09:42:59.524-0500 [DEBUG] provider.terraform-provider-gitlab_v3.8.0: -----------------------------------------------------
2022-01-03T09:42:59.526-0500 [DEBUG] provider.terraform-provider-gitlab_v3.8.0: 2022/01/03 09:42:59 [DEBUG] update gitlab project 600
2022-01-03T09:42:59.526-0500 [DEBUG] provider.terraform-provider-gitlab_v3.8.0: 2022/01/03 09:42:59 [DEBUG] GitLab API Request Details:
2022-01-03T09:42:59.526-0500 [DEBUG] provider.terraform-provider-gitlab_v3.8.0: ---[ REQUEST ]---------------------------------------
2022-01-03T09:42:59.526-0500 [DEBUG] provider.terraform-provider-gitlab_v3.8.0: PUT /api/v4/projects/600 HTTP/1.1
2022-01-03T09:42:59.526-0500 [DEBUG] provider.terraform-provider-gitlab_v3.8.0: Host: gitlab.mycorp.com
2022-01-03T09:42:59.526-0500 [DEBUG] provider.terraform-provider-gitlab_v3.8.0: User-Agent: HashiCorp Terraform/1.1.2 (+https://www.terraform.io) Terraform Plugin SDK/1.16.1 terraform-provider-gitlab
2022-01-03T09:42:59.526-0500 [DEBUG] provider.terraform-provider-gitlab_v3.8.0: Content-Length: 31
2022-01-03T09:42:59.526-0500 [DEBUG] provider.terraform-provider-gitlab_v3.8.0: Accept: application/json
2022-01-03T09:42:59.526-0500 [DEBUG] provider.terraform-provider-gitlab_v3.8.0: Authorization: Bearer P4zZBoxecEJeZXUeNF7y
2022-01-03T09:42:59.526-0500 [DEBUG] provider.terraform-provider-gitlab_v3.8.0: Content-Type: application/json
2022-01-03T09:42:59.526-0500 [DEBUG] provider.terraform-provider-gitlab_v3.8.0: Accept-Encoding: gzip
2022-01-03T09:42:59.526-0500 [DEBUG] provider.terraform-provider-gitlab_v3.8.0: 
2022-01-03T09:42:59.526-0500 [DEBUG] provider.terraform-provider-gitlab_v3.8.0: {
2022-01-03T09:42:59.526-0500 [DEBUG] provider.terraform-provider-gitlab_v3.8.0:  "squash_option": "default_off"
2022-01-03T09:42:59.526-0500 [DEBUG] provider.terraform-provider-gitlab_v3.8.0: }
2022-01-03T09:42:59.526-0500 [DEBUG] provider.terraform-provider-gitlab_v3.8.0: -----------------------------------------------------
```

Note the **PUT** - should it be **POST** instead?

```
2022-01-03T09:42:59.771-0500 [ERROR] vertex "module.genesis.gitlab_project.my_project[\"chart\"]" error: PUT https://gitlab.mycorp.com/api/v4/projects/600: 400 {error: allow_merge_on_skipped_pipeline, autoclose_referenced_issues, auto_devops_enabled, auto_devops_deploy_strategy, auto_cancel_pending_pipelines, build_coverage_regex, build_git_strategy, build_timeout, builds_access_level, ci_config_path, ci_default_git_depth, ci_forward_deployment_enabled, container_registry_enabled, container_expiration_policy_attributes, default_branch, description, emails_disabled, forking_access_level, issues_access_level, lfs_enabled, merge_requests_access_level, merge_method, name, only_allow_merge_if_all_discussions_are_resolved, only_allow_merge_if_pipeline_succeeds, pages_access_level, path, printing_merge_request_link_enabled, public_builds, remove_source_branch_after_merge, repository_access_level, request_access_enabled, resolve_outdated_diff_discussions, restrict_user_defined_variables, shared_runners_enabled, snippets_access_level, tag_list, visibility, wiki_access_level, avatar, suggestion_commit_message, repository_storage, compliance_framework_setting, packages_enabled, service_desk_enabled, issues_enabled, jobs_enabled, merge_requests_enabled, wiki_enabled, snippets_enabled, approvals_before_merge, external_authorization_classification_label, fallback_approvals_required, import_url, issues_template, merge_requests_template are missing, at least one parameter must be provided}
```

### Panic Output
N/A

### Expected Behavior
terraform should have sent a correct API request (probably POST instead of PUT).

### Actual Behavior
terraform sent an incorrect API request and received a 400 from GitLab.

### Steps to Reproduce
1. Upgrade provider from 3.7.0 to 3.8.0, but make no other change to a valid existing project.
2. `terraform apply`

### Important Factoids
None that I know of.

### References
- [Issue 770](https://github.com/gitlabhq/terraform-provider-gitlab/issues/770).

--- a/docs/resources/project.md
+++ b/docs/resources/project.md
@@ -78,7 +78,7 @@ resource "gitlab_project" "example-two" {
 - **request_access_enabled** (Boolean) Allow users to request member access.
 - **shared_runners_enabled** (Boolean) Enable shared runners for this project.
 - **snippets_enabled** (Boolean) Enable snippets for the project.
-- **squash_option** (String) Squash commits when merge request. Valid values are `never`, `always`, `default_on`, or `default_off`. The default value is `default_off`.
+- **squash_option** (String) Squash commits when merge request. Valid values are `never`, `always`, `default_on`, or `default_off`. The default value is `default_off`. [GitLab >= 14.1]
 - **tags** (Set of String) Tags (topics) of the project.
 - **template_name** (String) When used without use_custom_template, name of a built-in project template. When used with use_custom_template, name of a custom project template. This option is mutually exclusive with `template_project_id`.
 - **template_project_id** (Number) When used with use_custom_template, project ID of a custom project template. This is preferable to using template_name since template_name may be ambiguous (enterprise edition). This option is mutually exclusive with `template_name`.

--- a/internal/provider/resource_gitlab_project.go
+++ b/internal/provider/resource_gitlab_project.go
@@ -193,7 +193,7 @@ var resourceGitLabProjectSchema = map[string]*schema.Schema{
 		Optional:    true,
 	},
 	"squash_option": {
-		Description:  "Squash commits when merge request. Valid values are `never`, `always`, `default_on`, or `default_off`. The default value is `default_off`.",
+		Description:  "Squash commits when merge request. Valid values are `never`, `always`, `default_on`, or `default_off`. The default value is `default_off`. [GitLab >= 14.1]",
 		Type:         schema.TypeString,
 		Optional:     true,
 		Default:      "default_off",
@@ -386,7 +386,7 @@ var _ = registerResource("gitlab_project", func() *schema.Resource {
 	}
 })
 
-func resourceGitlabProjectSetToState(d *schema.ResourceData, project *gitlab.Project) error {
+func resourceGitlabProjectSetToState(client *gitlab.Client, d *schema.ResourceData, project *gitlab.Project) error {
 	d.SetId(fmt.Sprintf("%d", project.ID))
 	d.Set("name", project.Name)
 	d.Set("path", project.Path)
@@ -417,7 +417,11 @@ func resourceGitlabProjectSetToState(d *schema.ResourceData, project *gitlab.Pro
 		return err
 	}
 	d.Set("archived", project.Archived)
-	d.Set("squash_option", project.SquashOption)
+	if supportsSquashOption, err := isGitLabVersionAtLeast(client, "14.1")(); err != nil {
+		return err
+	} else if supportsSquashOption {
+		d.Set("squash_option", project.SquashOption)
+	}
 	d.Set("remove_source_branch_after_merge", project.RemoveSourceBranchAfterMerge)
 	d.Set("packages_enabled", project.PackagesEnabled)
 	d.Set("pages_access_level", string(project.PagesAccessLevel))
@@ -453,7 +457,6 @@ func resourceGitlabProjectCreate(ctx context.Context, d *schema.ResourceData, me
 		OnlyAllowMergeIfAllDiscussionsAreResolved: gitlab.Bool(d.Get("only_allow_merge_if_all_discussions_are_resolved").(bool)),
 		AllowMergeOnSkippedPipeline:               gitlab.Bool(d.Get("allow_merge_on_skipped_pipeline").(bool)),
 		SharedRunnersEnabled:                      gitlab.Bool(d.Get("shared_runners_enabled").(bool)),
-		SquashOption:                              stringToSquashOptionValue(d.Get("squash_option").(string)),
 		RemoveSourceBranchAfterMerge:              gitlab.Bool(d.Get("remove_source_branch_after_merge").(bool)),
 		PackagesEnabled:                           gitlab.Bool(d.Get("packages_enabled").(bool)),
 		Mirror:                                    gitlab.Bool(d.Get("mirror").(bool)),
@@ -513,6 +516,14 @@ func resourceGitlabProjectCreate(ctx context.Context, d *schema.ResourceData, me
 
 	if v, ok := d.GetOk("ci_config_path"); ok {
 		options.CIConfigPath = gitlab.String(v.(string))
+	}
+
+	if supportsSquashOption, err := isGitLabVersionAtLeast(client, "14.1")(); err != nil {
+		return diag.FromErr(err)
+	} else if supportsSquashOption {
+		if v, ok := d.GetOk("squash_option"); ok {
+			options.SquashOption = stringToSquashOptionValue(v.(string))
+		}
 	}
 
 	log.Printf("[DEBUG] create gitlab project %q", *options.Name)
@@ -672,7 +683,7 @@ func resourceGitlabProjectRead(ctx context.Context, d *schema.ResourceData, meta
 		return nil
 	}
 
-	if err := resourceGitlabProjectSetToState(d, project); err != nil {
+	if err := resourceGitlabProjectSetToState(client, d, project); err != nil {
 		return diag.FromErr(err)
 	}
 
@@ -782,7 +793,9 @@ func resourceGitlabProjectUpdate(ctx context.Context, d *schema.ResourceData, me
 		options.LFSEnabled = gitlab.Bool(d.Get("lfs_enabled").(bool))
 	}
 
-	if d.HasChange("squash_option") {
+	if supportsSquashOption, err := isGitLabVersionAtLeast(client, "14.1")(); err != nil {
+		return diag.FromErr(err)
+	} else if supportsSquashOption && d.HasChange("squash_option") {
 		options.SquashOption = stringToSquashOptionValue(d.Get("squash_option").(string))
 	}
 

--- a/internal/provider/resource_gitlab_project_test.go
+++ b/internal/provider/resource_gitlab_project_test.go
@@ -964,11 +964,11 @@ func testAccCheckAggregateGitlabProject(expected, received *gitlab.Project) reso
 				}
 			}
 
-			if err := resourceGitlabProjectSetToState(expectedData, expected); err != nil {
+			if err := resourceGitlabProjectSetToState(testGitlabClient, expectedData, expected); err != nil {
 				return err
 			}
 
-			if err := resourceGitlabProjectSetToState(receivedData, received); err != nil {
+			if err := resourceGitlabProjectSetToState(testGitlabClient, receivedData, received); err != nil {
 				return err
 			}
 

